wgpu-hal/src/vulkan/device.rs

```diff
@@ -1452,119 +1418,116 @@ impl crate::Device for super::Device {
+                            .buffer(binding.buffer.raw)
+                            .offset(binding.offset)
+                            .range(binding.size.map_or(vk::WHOLE_SIZE, wgt::BufferSize::get))
+                    }
+                    write.buffer_info(local_infos)
                 }
                 vk::DescriptorType::ACCELERATION_STRUCTURE_KHR => {
-                    let index = acceleration_structure_infos.len();
                     let start = entry.resource_index;
                     let end = start + entry.count;
 
-                    let raw_start = raw_acceleration_structures.len();
-
-                    raw_acceleration_structures.extend(
-                        desc.acceleration_structures[start as usize..end as usize]
-                            .iter()
-                            .map(|acceleration_structure| acceleration_structure.raw),
-                    );
+                    let local_raw_acceleration_structures;
+                    (
+                        local_raw_acceleration_structures,
+                        raw_acceleration_structures,
+                    ) = raw_acceleration_structures.split_at_mut(entry.count as usize);
 
-                    let acceleration_structure_info =
-                        vk::WriteDescriptorSetAccelerationStructureKHR::builder()
-                            .acceleration_structures(&raw_acceleration_structures[raw_start..]);
-
-                    // todo: Dereference the struct to get around lifetime issues. Safe as long as we never resize
-                    // `raw_acceleration_structures`.
-                    let acceleration_structure_info: vk::WriteDescriptorSetAccelerationStructureKHR = *acceleration_structure_info;
+                    for (local_raw_as, acceleration_structure) in local_raw_acceleration_structures
+                        .iter_mut()
+                        .zip(&desc.acceleration_structures[start as usize..end as usize])
+                    {
+                        *local_raw_as = acceleration_structure.raw;
+                    }
 
-                    assert!(
-                        index < desc.acceleration_structures.len(),
-                        "Encountered more acceleration structures then expected"
-                    );
-                    acceleration_structure_infos.push(acceleration_structure_info);
+                    let local;
+                    (local, acceleration_structure_infos) =
+                        acceleration_structure_infos.split_first_mut().unwrap();
 
-                    extra_descriptor_count += 1;
+                    *local = vk::WriteDescriptorSetAccelerationStructureKHR::default()
+                        .acceleration_structures(local_raw_acceleration_structures);
 
-                    write.push_next(&mut acceleration_structure_infos[index])
+                    write.descriptor_count(entry.count).push_next(local)
```

Comment:
Casual bug-fix here: `write.descriptor_count` should be equal to `acceleration_structure_count` in `vk::WriteDescriptorSetAccelerationStructureKHR`. Unless the selected range of `desc.acceleration_structures` (is `entry.count`) is always 1?

Reply:
Makes sense to me.

@daniel-keitel do you have more context from https://github.com/gfx-rs/wgpu/pull/3507?